clar2wasm/src/cost.rs

```diff
@@ -0,0 +1,743 @@
+}
+
+/// Cost assessment function
+#[derive(Debug, Clone, Copy)]
+enum Caf {
+    Constant(u32),
+    Linear { a: u64, b: u64 },
+    NLogN { a: u64, b: u64 },
+    None,
+}
+
+impl ChargeContext {
+    fn emit(&self, instrs: &mut InstrSeqBuilder, cost: &WordCost, n: impl Into<Scalar>) {
+        let n = n.into();
+
+        self.emit_with_caf(
+            instrs,
+            cost.runtime,
+            self.runtime,
+            ErrorMap::CostOverrunRuntime as _,
+            n,
+        );
+        self.emit_with_caf(
+            instrs,
+            cost.read_count,
+            self.read_count,
+            ErrorMap::CostOverrunReadCount as _,
+            n,
+        );
+        self.emit_with_caf(
+            instrs,
+            cost.read_length,
+            self.read_length,
+            ErrorMap::CostOverrunReadLength as _,
+            n,
+        );
+        self.emit_with_caf(
+            instrs,
+            cost.write_count,
+            self.write_count,
+            ErrorMap::CostOverrunWriteCount as _,
+            n,
+        );
+        self.emit_with_caf(
+            instrs,
+            cost.write_length,
+            self.write_length,
+            ErrorMap::CostOverrunWriteLength as _,
+            n,
+        );
+    }
+
+    fn emit_with_caf(
+        &self,
+        instrs: &mut InstrSeqBuilder,
+        params: Caf,
+        global: GlobalId,
+        err_code: i32,
+        n: impl Into<Scalar>,
+    ) {
+        match params {
+            Caf::Constant(n) => caf_const(instrs, global, self.runtime_error, err_code, n),
+            Caf::Linear { a, b } => {
+                caf_linear(instrs, global, self.runtime_error, err_code, n, a, b)
+            }
+            Caf::NLogN { a, b } => caf_nlogn(instrs, global, self.runtime_error, err_code, n, a, b),
+            Caf::None => {}
+        }
+    }
+}
+
+// NOTE: since we're only measuring costs during runtime - and everything
+//       else is left for the analyzer
+
+fn caf_const(
+    instrs: &mut InstrSeqBuilder,
+    global: GlobalId,
+    error: FunctionId,
+    err_code: i32,
+    cost: impl Into<Scalar>,
+) {
+    let cost = cost.into();
+
+    instrs.global_get(global).scalar_get(cost);
+
+    // global - cost
+    instrs
+        .instr(Instr::Binop(Binop {
+            op: BinaryOp::I64Sub,
+        }))
+        .global_set(global)
+        .global_get(global)
+        .i64_const(0)
+        .instr(Instr::Binop(Binop {
+            op: BinaryOp::I64LtS,
+        }))
+        .if_else(
+            None,
+            |builder| {
+                builder.i32_const(err_code);
+                builder.call(error);
+            },
+            |_| {},
+        );
+}
+
+fn caf_linear(
+    instrs: &mut InstrSeqBuilder,
+    global: GlobalId,
+    error: FunctionId,
+    err_code: i32,
+    n: impl Into<Scalar>,
+    a: u64,
+    b: u64,
+) {
+    let n = n.into();
+
+    // cost = (+ (* a n) b))
+    instrs
+        .global_get(global)
```

Comment:
Why is the global here?

Reply:
The cost assessment functions code was a little bit unclear, so I changed it around a little according to some of your recommendations below as well. Unfortunately, since subtraction is non-commutative, we have to push the global before the cost, so it remains in the same place.